paddle/fluid/framework/details/analysis_var_pass.cc

```diff
@@ -0,0 +1,280 @@
+// Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "paddle/fluid/framework/details/analysis_var_pass.h"
+#include <algorithm>
+#include <atomic>
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <memory>
+#include <sstream>
+#include <string>
+#include <type_traits>
+#include <vector>
+#include "gflags/gflags.h"
+#include "paddle/fluid/framework/data_type.h"
+
+DEFINE_bool(enable_subgraph_optimize, false,
+            "SubGraph also reuse global graph variables, it will reduce the "
+            "memory occupation"
+            "but a higher risk of memory reuse error. default disabled.");
+DEFINE_string(memory_optimize_debug, "",
+              "debug the operator output variable when do the variable reuse."
+              "memory reuse pass."
+              "only for debug, default disabled.");
+
+namespace paddle {
+namespace framework {
+namespace details {
+
+std::unordered_set<ir::Node*> AnalysisVarPass::GetSubBlockOutputVars(
+    const std::unordered_set<ir::Node*>& nodes) const {
+  std::unordered_set<ir::Node*> vars;
+  std::unordered_map<std::string, ir::Node*> var_to_node_map;
+  ProgramDesc* program = nullptr;
+  for (auto& op : nodes) {
+    if (op->IsOp() && program == nullptr) {
+      program = op->Op()->Block()->Program();
+    }
+    if (op->IsVar() && !op->IsCtrlVar()) {
+      var_to_node_map[op->Name()] = op;
+    }
+  }
+  if (program->Size() > 1) {
+    // size > 1 means program has subblock. A subblock's AllVars
+    // only contains the output variables, no input variables
```

Comment:
Really. If in the while_op's block, I create a variable that is only used on while_op, it will show up in global block? I think it should appear in local block